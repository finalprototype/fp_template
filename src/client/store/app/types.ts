import keyMirror from '../../utils/keyMirror';

export default keyMirror([
  'APP_INIT_CONFIG',
]);

export interface AppState {
  isReady: boolean;
  config: AppConfig;
}

// Interfaces
export interface AppConfig {
  env: string;
  version: string;
  assetsPath: string;
  manifest: { [key: string]: string|string[] };
  flags: string[];
}

export interface AppAction {
  type: string;
  payload?: { [key: string]: string|undefined };
}
