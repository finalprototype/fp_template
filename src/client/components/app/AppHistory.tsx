import { createBrowserHistory, History } from 'history';

export const history = createBrowserHistory<History>();
