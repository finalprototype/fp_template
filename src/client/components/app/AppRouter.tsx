import React, { lazy } from 'react';
import { Route, Redirect, Switch } from 'react-router-dom';

const HomeContainer = lazy(() => import('../home/HomeContainer'));

const AppRouter: React.FunctionComponent = () => (
  <Switch>
    <Route path="/">
      <HomeContainer />
    </Route>
    <Redirect from="*" to="/" />
  </Switch>
);

export default AppRouter;
