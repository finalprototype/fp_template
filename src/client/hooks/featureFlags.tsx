import qs from 'qs';
import { useLocation } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { uniq, difference } from 'lodash';

import { RootState } from '../store';

export const useActiveFlags = (): string[] => {
  const { config } = useSelector((state: RootState) => state.app);
  const initialflags = config.flags;
  const { search } = useLocation();
  const {
    ffe = '',
    ffd = '',
  } = qs.parse(search, {
    ignoreQueryPrefix: true,
  });

  const urlFlagsEnabled = ffe.length
    ? (ffe as string).split(',')
    : [];
  const urlFlagsDisabled = ffd && ffd.length
    ? (ffd as string).split(',')
    : [];

  const updatedFlags = difference(
    uniq(initialflags.concat(urlFlagsEnabled)),
    urlFlagsDisabled
  );

  return updatedFlags;
};
