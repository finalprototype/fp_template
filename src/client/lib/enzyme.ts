import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({
  adapter: new Adapter(),
  disableLifecycleMethods: true,
});

export const shallow = Enzyme.shallow;
export const mount = Enzyme.mount;
export const render = Enzyme.render;

export default Enzyme;
