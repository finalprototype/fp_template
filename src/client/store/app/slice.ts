import { createSlice } from '@reduxjs/toolkit';

import { AppState } from './types';

export const getInitialState = (): AppState => ({
  isReady: false,
  config: {
    env: '',
    version: '',
    assetsPath: '',
    manifest: {},
    flags: [],
  },
});

const appSlice = createSlice({
  name: 'app',
  initialState: getInitialState(),
  reducers: {
    init(state, action) {
      state.config = { ...state.config, ...action.payload };
      state.isReady = true;
    },
  },
});

export const { init } = appSlice.actions;

export default appSlice.reducer;
