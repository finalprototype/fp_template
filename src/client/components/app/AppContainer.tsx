import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { history } from './AppHistory';
import AppView from './AppView';
import store from '../../store';

const AppContainer: React.FunctionComponent = () => (
  <Provider store={store}>
    <Router history={history!}>
      <AppView />
    </Router>
  </Provider>
);

export default AppContainer;
