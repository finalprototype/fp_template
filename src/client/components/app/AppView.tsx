import React, { Suspense, useEffect } from 'react';
import { connect, ConnectedProps } from 'react-redux';

import * as actions from '../../store/app/slice';
import Loader from '../ui/Loader';
import AppRouter from './AppRouter';

import styles from './__styles__/AppView.scss';

const mapDispatch = { ...actions };
const connector = connect(null, mapDispatch);
type PropsFromRedux = ConnectedProps<typeof connector>;

const AppView: React.FunctionComponent<PropsFromRedux> = ({ init }: PropsFromRedux) => {
  useEffect(() => {
    if (!window.config) return;
    const { config } = window;
    init({
      env: config.env,
      version: config.version,
      assetsPath: config.assets_path,
      manifest: { ...config.manifest },
      flags: [...config.flags],
    });
    delete window.config;
  }, [init]);

  const suspenseFallback = (
    <Loader className={styles['suspense-loader']} />
  );

  return (
    <div className={styles.content}>
      <Suspense fallback={suspenseFallback}>
        <AppRouter />
      </Suspense>
    </div>
  );
};

export default connector(AppView);
