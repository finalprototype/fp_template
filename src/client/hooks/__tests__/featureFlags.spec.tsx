import React from 'react';
import qs from 'qs';
import { shallow } from 'enzyme';

import { AppState } from '../../store/app/types';

import { useActiveFlags } from '../featureFlags';

let mockLocation = '/path';
let localConfig = {};
let paramsObj: { [key: string]: string|number|boolean|undefined } = {};
const mockAppConfig = {
  env: 'test',
  version: '0',
  assetsPath: '',
  manifest: {},
  flags: [],
};
const setUrlFlags = (enabled: boolean, ...flags: string[]): void => {
  const key = enabled ? 'ffe' : 'ffd';
  paramsObj[key] = flags.join(',');
};
const mockLocationReturn = (): { [key: string]: string } => ({
  pathname: mockLocation,
  search: qs.stringify(paramsObj, {
    addQueryPrefix: true,
    format: 'RFC1738',
  }),
});
const mockAppState = (): AppState => ({
  isReady: true,
  config: {
    ...mockAppConfig,
    ...localConfig,
  },
});

jest.mock('react-redux', () => ({
  // @ts-ignore
  ...jest.requireActual('react-redux'),
  useSelector: () => mockAppState(),
}));

jest.mock('react-router-dom', () => ({
  // @ts-ignore
  ...jest.requireActual('react-router-dom'),
  useLocation: () => mockLocationReturn(),
}));

describe('hooks/featureFlags', () => {
  interface HookWrapperProps extends React.HTMLAttributes<HTMLDivElement> {
    hook: () => string[];
  }
  // @ts-ignore
  const HookWrapper = ({ hook }: HookWrapperProps) => <div hook={hook()} />;
  const cleanUp = () => {
    localConfig = {};
    mockLocation = '/path';
    paramsObj = {
      adummy: true,
      zdummmy: false,
    };
  };

  beforeEach(cleanUp);
  afterEach(cleanUp);

  describe('useActiveFlags', () => {
    const mockInitialFlags = [
      'initial1',
      'initial.2',
      'initial-3',
      'initial_4',
      'initial 5',
    ];
    const mockUrlEnabledFlags = [
      'url1',
      'url.2',
      'url-3',
      'url_4',
      'url 5',
    ];
    const mockSharedFlags = [
      'shared1',
      'shared.2',
      'shared-3',
      'shared_4',
      'shared 5',
    ];

    it('returns empty flags', () => {
      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns only initial flags', () => {
      localConfig = { flags: mockInitialFlags };

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns only url enabled flags', () => {
      setUrlFlags(
        true,
        ...mockUrlEnabledFlags
      );

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns unique initial and url enabled flags', () => {
      localConfig = {
        flags: mockInitialFlags.concat(mockSharedFlags),
      };
      setUrlFlags(
        true,
        ...mockUrlEnabledFlags.concat(mockSharedFlags),
      );

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns without url disabled flags', () => {
      const disabledFlags = [
        'initial1',
        'shared.2',
        'url-3',
        'initial_4',
        'shared 5',
      ];

      localConfig = {
        flags: mockInitialFlags.concat(mockSharedFlags),
      };
      setUrlFlags(
        true,
        ...mockUrlEnabledFlags.concat(mockSharedFlags),
      );
      setUrlFlags(
        false,
        ...disabledFlags,
      );

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook).toEqual(
        expect.not.arrayContaining(disabledFlags)
      );
      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns empty if all flags disabled', () => {
      localConfig = {
        flags: mockInitialFlags.concat(mockSharedFlags),
      };
      setUrlFlags(
        true,
        ...mockUrlEnabledFlags.concat(mockSharedFlags),
      );
      setUrlFlags(
        false,
        ...mockInitialFlags,
        ...mockSharedFlags,
        ...mockUrlEnabledFlags,
      );

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook.length).toEqual(0);
      expect(wrapper.props().hook).toMatchSnapshot();
    });

    it('returns empty safely if only flags disabled', () => {
      setUrlFlags(
        false,
        ...mockInitialFlags,
        ...mockSharedFlags,
        ...mockUrlEnabledFlags,
      );

      const wrapper = shallow(
        <HookWrapper hook={() => useActiveFlags()} />
      );

      expect(wrapper.props().hook.length).toEqual(0);
      expect(wrapper.props().hook).toMatchSnapshot();
    });
  });
});
