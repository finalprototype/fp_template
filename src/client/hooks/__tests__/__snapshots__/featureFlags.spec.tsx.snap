// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks/featureFlags useActiveFlags returns empty flags 1`] = `Array []`;

exports[`hooks/featureFlags useActiveFlags returns empty if all flags disabled 1`] = `Array []`;

exports[`hooks/featureFlags useActiveFlags returns empty safely if only flags disabled 1`] = `Array []`;

exports[`hooks/featureFlags useActiveFlags returns only initial flags 1`] = `
Array [
  "initial1",
  "initial.2",
  "initial-3",
  "initial_4",
  "initial 5",
]
`;

exports[`hooks/featureFlags useActiveFlags returns only url enabled flags 1`] = `
Array [
  "url1",
  "url.2",
  "url-3",
  "url_4",
  "url 5",
]
`;

exports[`hooks/featureFlags useActiveFlags returns unique initial and url enabled flags 1`] = `
Array [
  "initial1",
  "initial.2",
  "initial-3",
  "initial_4",
  "initial 5",
  "shared1",
  "shared.2",
  "shared-3",
  "shared_4",
  "shared 5",
  "url1",
  "url.2",
  "url-3",
  "url_4",
  "url 5",
]
`;

exports[`hooks/featureFlags useActiveFlags returns without url disabled flags 1`] = `
Array [
  "initial.2",
  "initial-3",
  "initial 5",
  "shared1",
  "shared-3",
  "shared_4",
  "url1",
  "url.2",
  "url_4",
  "url 5",
]
`;
